atlasdb-commons/src/test/java/com/palantir/async/initializer/AsyncInitializerTest.java

```diff
@@ -48,38 +56,109 @@ protected int sleepIntervalInMillis() {
+                .hasMessageContaining("Multiple calls tried to initialize the same instance.");
+        initializer.deterministicScheduler.tick(0, TimeUnit.SECONDS);
+        assertThat(initializer.initializationAttempts).isEqualTo(1);
     }
 
     @Test
-    public void initializationIsNoopIfAlreadyInitialized() {
-        AlwaysFailingInitializer initializer = new AlwaysFailingInitializer() {
+    public void asyncInitializationKeepsRetryingAndEventuallySucceeds() throws InterruptedException {
+        AlwaysFailingInitializer eventuallySuccessfulInitializer = new AlwaysFailingInitializer() {
+            @Override
+            public void tryInitialize() {
+                if (initializationAttempts < 5) {
+                    super.tryInitialize();
+                }
+            }
+        };
+
+        eventuallySuccessfulInitializer.initialize(true);
+        assertFalse(eventuallySuccessfulInitializer.isInitialized());
+        eventuallySuccessfulInitializer.deterministicScheduler.tick(5 * ASYNC_INIT_DELAY + 1, TimeUnit.MILLISECONDS);
+        assertThat(eventuallySuccessfulInitializer.isInitialized()).isTrue();
+    }
+
+    @Test
+    public void canCancelInitializationAndPerformCleanupIfInitialized() throws InterruptedException {
+        AlwaysFailingInitializer initializer = new AlwaysFailingInitializer();
+        AlwaysFailingInitializer successfulInitializer = new AlwaysFailingInitializer() {
             @Override
             public boolean isInitialized() {
                 return true;
             }
         };
+        Runnable cleanupTask = mock(Runnable.class);
+        doNothing().when(cleanupTask).run();
+
+        initializeAsyncCancelAndVerifyCancelled(initializer, cleanupTask);
+        verify(cleanupTask, never()).run();
+
+        initializeAsyncCancelAndVerifyCancelled(successfulInitializer, cleanupTask);
+        verify(cleanupTask).run();
+    }
+
+    private AsyncInitializer getMockedInitializer() {
+        AsyncInitializer initializer = mock(AsyncInitializer.class, Mockito.CALLS_REAL_METHODS);
+        doNothing().when(initializer).assertNeverCalledInitialize();
+        doThrow(new RuntimeException("Failed initializing")).when(initializer).tryInitialize();
+        doNothing().when(initializer).scheduleInitialization();
+        return initializer;
+    }
+
+    private void initializeAsyncCancelAndVerifyCancelled(AlwaysFailingInitializer initializer,
+            Runnable cleanupTask) throws InterruptedException {
         initializer.initialize(true);
-        assertThat(initializer.initializationAttempts).isEqualTo(0);
-        initializer.initialize(false);
-        assertThat(initializer.initializationAttempts).isEqualTo(0);
+        initializer.cancelInitialization(cleanupTask);
+        int numberOfAttemptsWhenCancelled = initializer.initializationAttempts;
+        initializer.deterministicScheduler.tick(ASYNC_INIT_DELAY * 5 + 1, TimeUnit.MILLISECONDS);
```

Comment:
We have `ASYNC_INIT_DELAY * 5 + 1` twice here (well, the other time, it's `5*AID+1` 🙂  ).
Is it worth extracting a constant here, like `SEVERAL_INIT_DELAYS`?